File: app/src/main/java/com/skydoves/waterdays/ui/viewholders/DailyDrinkViewHolder.java
package com.skydoves.waterdays.ui.viewholders;

import android.content.DialogInterface;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.view.View;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import com.skydoves.waterdays.R;
import com.skydoves.waterdays.models.Drink;

import butterknife.BindView;

/**
 * Developed by skydoves on 2017-08-20.
 * Copyright (c) 2017 skydoves rights reserved.
 */

public class DailyDrinkViewHolder extends BaseViewHolder {

    protected @BindView(R.id.item_dailyrecord_tv_todaydate) TextView tv_item_date;
    protected @BindView(R.id.item_dailyrecord_tv_drinkamount) TextView tv_item_amount;
    protected @BindView(R.id.item_dailyrecord_imv_main) ImageView imv_item_mainicon;
    protected @BindView(R.id.item_dailyrecord_btn_delete) ImageButton ibtn_item_delete;

    private Drink drink;

    private Delegate delegate;

    public interface Delegate {
        void onConfirm(Drink drink);
    }

    public DailyDrinkViewHolder(final @NonNull View view, final @NonNull Delegate delegate) {
        super(view);
        this.delegate = delegate;
    }

    @Override
    public void bindData(@NonNull Object data) throws Exception {
        this.drink = (Drink) data;

        tv_item_date.setText(drink.getDate());
        tv_item_amount.setText(drink.getAmount());
        imv_item_mainicon.setImageDrawable(drink.getImage());
        ibtn_item_delete.setOnClickListener(view -> {
            AlertDialog.Builder alertDlg = new AlertDialog.Builder(view.getContext());
            alertDlg.setTitle("알람");

            // Yes - delete
            alertDlg.setPositiveButton("예", (DialogInterface dialog, int which) -> {
                delegate.onConfirm(drink);
            });

            // No - cancel
            alertDlg.setNegativeButton("아니오", (DialogInterface dialog, int which) -> dialog.dismiss());
            alertDlg.setMessage(String.format("해당 기록을 지우시겠습니까?"));
            alertDlg.show();
        });
    }

    @Override
    public void onClick(View v) {

    }
}

File: app/src/main/java/com/skydoves/waterdays/ui/viewholders/DailyDrinkViewHolder.kt
/*
 * Copyright (C) 2016 skydoves
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.skydoves.waterdays.ui.viewholders

import android.content.DialogInterface
import android.view.View
import androidx.appcompat.app.AlertDialog
import com.skydoves.waterdays.models.Drink
import kotlinx.android.synthetic.main.item_dailyrecord.view.*

/**
 * Developed by skydoves on 2017-08-20.
 * Copyright (c) 2017 skydoves rights reserved.
 */

class DailyDrinkViewHolder(view: View, private val delegate: Delegate) : BaseViewHolder(view) {

  private lateinit var drink: Drink

  interface Delegate {
    fun onClick(view: View, drink: Drink)
    fun onConfirm(drink: Drink)
  }

  @Throws(Exception::class)
  override fun bindData(data: Any) {
    this.drink = data as Drink

    itemView.item_dailyrecord_tv_todaydate.setText(drink.date)
    itemView.item_dailyrecord_tv_drinkamount.text = drink.amount
    itemView.item_dailyrecord_imv_main.setImageDrawable(drink.image)
    itemView.item_dailyrecord_btn_delete.setOnClickListener {
      val alertDlg = AlertDialog.Builder(it.context)
      alertDlg.setTitle("알람")

      // yes - delete
      alertDlg.setPositiveButton("예") { dialog: DialogInterface, which: Int -> delegate.onConfirm(drink) }

      // no - cancel
      alertDlg.setNegativeButton("아니오") { dialog: DialogInterface, which: Int -> dialog.dismiss() }
      alertDlg.setMessage(String.format("해당 기록을 지우시겠습니까?"))
      alertDlg.show()
    }
  }

  override fun onClick(v: View) {
    delegate.onClick(this.view(), drink)
  }

  override fun onLongClick(v: View): Boolean {
    return false
  }
}
